src/os/bluestore/BlueStore.cc

```diff
@@ -8959,21 +8974,32 @@ int BlueStore::_rename(TransContext *txc,
   txc->t->rmkey(PREFIX_OBJ, oldo->key);
 
   // rewrite shards
-  oldo->extent_map.fault_range(db, 0, oldo->onode.size);
-  get_object_key(cct, new_oid, &new_okey);
-  for (auto &s : oldo->extent_map.shards) {
+  {
+    oldo->extent_map.fault_range(db, 0, oldo->onode.size);
+    bool first_key = true;
     string key;
-    get_extent_shard_key(oldo->key, s.offset, &key);
-    txc->t->rmkey(PREFIX_OBJ, key);
-    s.dirty = true;
+    for (auto &s : oldo->extent_map.shards) {
+      if (first_key) {
+        get_extent_shard_key(oldo->key, s.offset, &key);
+        first_key = false;
+      } else {
+        rewrite_extent_shard_key(s.offset, &key);
+      }
+      txc->t->rmkey(PREFIX_OBJ, key);
+      s.dirty = true;
```

Comment:
I have a weak suspicious that there is no need to set dirty here as it indicates that shard encoding has changed. And that's not true here. @liewegas - am I right? Hence single enumeration function from the above comment might be a bit simpler - no need to have tricks for handling both Shard and dirty_shard_t types.